### PR TITLE
Use mapping computed by GumTree to find previous element

### PR DESCRIPTION
**Pull request by slarse**
Fix #23

Adds the `SpoonMapping` class that allows us to fetch the mapped Spoon elements across prev/new ASTs. This fixes the problems of not finding the previous element for inserts.

Unfortunately, I had to do some major shuffling around in other places as well to get this to work. In particular, I had to configure a custom pretty-printer. The default one used in `CtCompilationUnit.prettyPrint()` is the auto-import printer, which also prints implicit elements (such as `this`).

I also had to remove import statements from some test files, because diffmin currently does not handle import statements (it relied on the auto-import pretty-printer to do it).

So, yes, unfortunately this PR is not quite as focused as one may like it to be. If you'd like me to split this into several PRs, then I can do that, but it'll have to wait until next week.

### CODE REVIEW

File: src/test/resources/update/invocation/new.java [hunk 1]
```diff
@@ -1,7 +1,5 @@
 package com.diffmin;
 
-import java.util.List;
```
[the remark this answers is not in the document]
Comment (algomaster99) [reply src/test/resources/update/invocation/new.java:3]: @slarse 
> With the import statements in there, tests failed when using a custom pretty-printer.

I had one question. The test failed because the auto-importer removed the unnecessary import statement at it was not being used anywhere?

File: src/main/java/com/diffmin/App.java [hunk 2]
```diff
@@ -218,13 +225,14 @@ public static void main(String[] args) {
...
+            App app = new App();
+            Pair<Diff, CtModel> diffAndModel =
+                    App.computeDiff(new File(args[0]), new File(args[1]));
+            app.generatePatch(diffAndModel.getFirst());
             app.applyPatch();
-            CtModel patchedCtModel = app.modelToBeModified;
+            CtModel patchedCtModel = diffAndModel.getSecond();
             System.out.println(app.displayModifiedModel(patchedCtModel));
-        } catch (Exception e) {
+        } catch (java.lang.Exception e) {
```
[the remark this answers is not in the document]
Comment (slarse) [reply src/main/java/com/diffmin/App.java]: This comment completely passed me by. It's sometimes rather tricky to figure out _what_ a GitHub notification refers to...

And yes, this happened because I used Spork, which internally uses Spoon. And there are a ton of corner cases in Spoon when resolving the package of a type, and sometimes it fails to mark it as implicit (and thus, it's printed, like it was here).

Anyway, you can find a lot of test files in the Spork repository, under `src/test/resources/clean/both_modified`. See the README for a quickstart: https://github.com/KTH/spork

File: src/test/resources/update/invocation/new.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (slarse) [reply src/test/resources/update/invocation/new.java:3]: Yep.

**Comment by algomaster99** [comment]
@slarse I remember you authored this class to find the mapping of the corresponding spoon element for Spork and then we made a copy over here because it was used here as well. Today, I discovered that there is a [class](https://github.com/SpoonLabs/gumtree-spoon-ast-diff/blob/master/src/main/java/gumtree/spoon/diff/support/SpoonSupport.java) in `gumtree-spoon` that already found the spoon mapping. I was wondering if you ever came across it before you authored `SpoonMapping`. If yes, do you remember the drawbacks in `SpoonSupport` that led you to write your own class?
